### PR TITLE
Reduce the number of event record calls

### PR DESCRIPTION
**Pull request by tbennun**
When two streams are the same, there is no wait event call. With this PR, there will neither be an event record call.

### CODE REVIEW

**Review by ndryden — APPROVED** [review]
Looks good to me, but note this may not be in line with the H coding style.

File: include/hydrogen/SynchronizeAPI.hpp [hunk 1]
```diff
@@ -3,39 +3,55 @@
...
+                             SyncInfo<D2> const &other,
+                             SyncInfo<Ds> const &...others) {
+  if constexpr (D == Device::GPU && D == D2) {
+    // When the streams are the same, there is no need to create
+    // synchronization points. Skip "other" call recursively with the rest.
+    if (master.Stream() == other.Stream()) {
+      AddSynchronizationPoint(master, others...);
+      return;
+    }
+  }
```
Comment (benson31) [inline include/hydrogen/SynchronizeAPI.hpp]:
```suggestion
#ifdef HYDROGEN_HAVE_GPU
  if constexpr (D == Device::GPU && D == D2) {
    // When the streams are the same, there is no need to create
    // synchronization points. Skip "other" call recursively with the rest.
    if (master.Stream() == other.Stream()) {
      AddSynchronizationPoint(master, others...);
      return;
    }
  }
#endif // HYDROGEN_HAVE_GPU
```

File: include/hydrogen/SynchronizeAPI.hpp [hunk 2]
```diff
@@ -3,39 +3,55 @@
...
+template <Device D, Device D2>
+void AddSynchronizationPoint(SyncInfo<D> const &master,
+                             SyncInfo<D2> const &other) {
+  if constexpr (D == Device::GPU && D == D2) {
+    // When the two streams are the same, there is no need to create
+    // synchronization points.
+    if (master.Stream() == other.Stream()) {
+      return;
+    }
+  }
```
Comment (benson31) [inline include/hydrogen/SynchronizeAPI.hpp]:
```suggestion
#ifdef HYDROGEN_HAVE_GPU
  if constexpr (D == Device::GPU && D == D2) {
    // When the two streams are the same, there is no need to create
    // synchronization points.
    if (master.Stream() == other.Stream()) {
      return;
    }
  }
#endif // HYDROGEN_HAVE_GPU
```

**Review by benson31 — APPROVED** [review]
I think this is ok, as long as the compiler doesn't think too hard about the body of an `if constexpr (false)` block.

**Pull request merged**